baybe/kernels/base.py

```diff
@@ -12,31 +14,60 @@
     unstructure_base,
 )
 from baybe.serialization.mixin import SerialMixin
-from baybe.utils.basic import filter_attributes
+from baybe.utils.basic import filter_attributes, get_parent_classes
+
+if TYPE_CHECKING:
+    import torch
 
 
 @define(frozen=True)
 class Kernel(ABC, SerialMixin):
     """Abstract base class for all kernels."""
 
-    lengthscale_prior: Optional[Prior] = field(default=None, kw_only=True)
-    """An optional prior on the kernel lengthscale."""
-
-    def to_gpytorch(self, *args, **kwargs):
+    def to_gpytorch(
+        self,
+        *,
+        ard_num_dims: Optional[int] = None,
+        batch_shape: Optional[torch.Size] = None,
+        active_dims: Optional[tuple[int, ...]] = None,
+    ):
         """Create the gpytorch representation of the kernel."""
         import gpytorch.kernels
 
+        # Fetch the necessary gpytorch constructor parameters of the kernel.
+        # NOTE: In gpytorch, some attributes (like the kernel lengthscale) are handled
+        # via the `gpytorch.kernels.Kernel` base class. Hence, it is not sufficient to
+        # just check the fields of the actual class, but also those of the base class.
         kernel_cls = getattr(gpytorch.kernels, self.__class__.__name__)
-        fields_dict = filter_attributes(object=self, callable_=kernel_cls.__init__)
+        parent_classes = get_parent_classes(kernel_cls)
+        fields_dict = {}
+        for parent_class in parent_classes:
```

Comment:
am I missing it or would this just look at base clases and ignore the `kernel_cls`?

Reply:
No, the function has an option to include the class itself which is set to `True` by default. I made it more explicit here.

Reply:
please can we change this? I know why you introduced it, but it now make a semantically questionable thing and is inconsistent with `get_subclasses`, also @AdrianSosic 

1. I would not consider a class a base of itself, hence a utility `get_base_classes` does not have to return itself, not even optionally
2. `get_subclasses` has no such option, but its simply the same as `get_base_classes` but in the other direction. These functions should be as consistent as possible
3. If the original class itself is desired it can simply be added at the point where the utility is called - for that reason Id probably revert the change making the return a set and not a list, but uniqueness-ensuring can be kept imo
4. the naming `get_base_classes` is not consistent with `get_subclasses`, one of them should change

Reply:
Can change that again. But what is the problem with the function returning a set instead of a list? If we do not return a set, we need to either have an explicit check whether a class is already contained in the list (which happens for sure if we have multiple paths of base classes since they e.g. all end up at `object` in the end)

Reply:
Changed it such that it basically replicates the `get_subclasses` method, the only difference is now a check whether the class is already contained in the list to ensure uniqueness. I also renamed it accordingly, but did not amend the commit yet in case there are more changes required/desired by either one of you.

Reply:
thread has become obsolete

Reply:
i think this design pattern would be nice: `get_baseclasses(...) + [my_custom_addition1, my_custom_addition2]` which has direct use in the case that you want to add the class itself like `get_baseclasses(...) + [self.__class__]`

Reply:
is this current line even correct?
`for parent_class in [kernel_cls, base_classes]:`, this resolves to `for parent_class in [kernel_cls, {baseclass1, baseclass2, ...}]:` which doesnt seem right? `*` missing

Reply:
Veeeeery good catch, indeed! Must have slipped through Alex's debugging and I didn't check carefully enough! Fixed it. Could you please double-check an resolve if OK? Plus: accept if you are happy with then PR, then I can merge

Reply:
is this maybe where the mypy misc error (currently ignored) in that section stems from? If so please remove, 
resolving here for now

Reply:
In fact turned out to be related somehow. Still don't know what exactly caused it, but it's somehow related to the recent changes... Anyway, ignore is no longer needed, so I removed it :D Great that you noticed 🎖️